utils/sharedcache/sharedcache_immutable.go

```diff
@@ -0,0 +1,232 @@
+package sharedcache
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/ARM-software/golang-utils/utils/commonerrors"
+	"github.com/ARM-software/golang-utils/utils/filesystem"
+	"github.com/ARM-software/golang-utils/utils/idgen"
+	"github.com/ARM-software/golang-utils/utils/parallelisation"
+)
+
+const (
+	partFileDescriptor     = ".part"
+	defaultCachedPackageID = "cb93fdbe-6c2e-4f7d-96ac-c422fc52618e "
+)
+
+type SharedImmutableCacheRepository struct {
+	AbstractSharedCacheRepository
+}
+
+type FileWithModTime struct {
+	filename string
+	modTime  time.Time
+}
+
+func NewSharedImmutableCacheRepository(cfg *Configuration, fs filesystem.FS) (repository *SharedImmutableCacheRepository, err error) {
+	abstractCache, err := NewAbstractSharedCacheRepository(cfg, fs)
+	if err != nil {
+		return
+	}
+	repository = &SharedImmutableCacheRepository{
+		AbstractSharedCacheRepository: *abstractCache,
+	}
+	return
+}
+
+func listCompleteFilesByModTime(ctx context.Context, fs filesystem.FS, entryDir string) (sorted []string, err error) {
+	err = parallelisation.DetermineContextError(ctx)
+	if err != nil {
+		return
+	}
+
+	// sort the files by mod time
+	var fileModTimes []FileWithModTime
+	files, err := fs.Ls(entryDir)
+	if err != nil {
+		return
+	}
+
+	// Create array of non .part and non .hash files with their modtimes
+	for _, file := range files {
+		err = parallelisation.DetermineContextError(ctx)
+		if err != nil {
+			return sorted, err
+		}
+		isPartFile := strings.EqualFold(filepath.Ext(file), partFileDescriptor)
+		isHashFile := strings.EqualFold(filepath.Ext(file), hashFileDescriptor)
+		if !isPartFile && !isHashFile {
+			fullPath := filepath.Join(entryDir, file)
+			statInfo, err := fs.StatTimes(fullPath)
+			if err != nil {
+				return sorted, err
+			}
+			fileModTime := statInfo.ModTime()
+			fileModTimes = append(fileModTimes, FileWithModTime{
+				filename: file,
+				modTime:  fileModTime,
+			})
+		}
+	}
+	// can't do this directly using strings and stattimes to check on the
+	// file because the sort.Slice requires the function to output bool
+	// so if we wouldn't be abele to return an error if stattimes failed
+	sort.Slice(fileModTimes, func(i, j int) bool { return fileModTimes[i].modTime.After(fileModTimes[j].modTime) })
+
+	// map to just string
+	for _, file := range fileModTimes {
```

Comment:
You told me a while ago that we should do:
```go
for i := range xxx {
   y = xxx[i]
   ...
}
```
Not:
```go
for _, y := range xxx {
   ...
}
```
Is this still the case? If so then this look should be changed to use the index

Reply:
good spot